cstar/roms/runtime_settings.py

```diff
@@ -0,0 +1,806 @@
+            line = line.strip()
+            if not line or line.startswith("!"):
+                continue
+
+            if ":" in line:
+                # save the previous section if one was open
+                if current_section is not None:
+                    sections[current_section] = section_lines
+
+                # start a new section
+                current_section = line.split(":", 1)[0].strip()
+                section_lines = []
+            else:
+                section_lines.append(line)
+
+        # save the last section
+        if current_section is not None:
+            sections[current_section] = section_lines
+
+        return sections
+
+    @classmethod
+    def from_file(cls, filepath: PathLike[str]) -> "ROMSRuntimeSettings":
+        """Read ROMS runtime settings from a `.in` file.
+
+        ROMS runtime settings are specified via a `.in` file with sections corresponding
+        to different sets of parameters (e.g. time_stepping, bottom_drag). This method
+        first parses the file, creating a dict[str,list[str]] where the keys are the
+        section names and the values are a list of lines in that section. It then
+        translates that dictionary into a ROMSRuntimeSettings instance with properly
+        formatted attributes.
+
+        Parameters
+        ----------
+        - filepath (Path or str):
+           The path to the `.in` file
+
+        See Also
+        --------
+        - `ROMSRuntimeSettings.to_file()`: writes a ROMSRuntimeSettings instance to a
+           ROMS-compatible `.in` file
+        """
+
+        # Read file
+        filepath = Path(filepath)
+        sections = cls._load_raw_sections(filepath)
+        required_fields = {"title", "time_stepping", "bottom_drag", "output_root_name"}
+        missing_required_fields = required_fields - sections.keys()
+        if missing_required_fields:
+            raise ValueError(
+                "Required field missing from file. Required fields: "
+                "\n- title"
+                "\n- time_stepping"
+                "\n- bottom_drag"
+                "\n- output_root_name"
+            )
+        return cls(**sections)
+
+    def __str__(self) -> str:
+        """Returns a string representation of the input settings.
+
+        Returns
+        -------
+        str
+            A formatted string summarizing the ROMS input settings.
+        """
+        class_name = self.__class__.__name__
+        lines = [f"{class_name}"]
+        lines.append("-" * len(class_name))
+        lines.append(f"Title (`ROMSRuntimeSettings.title`): {self.title}")
+        lines.append(
+            f"Output filename prefix (`ROMSRuntimeSettings.output_root_name`): {self.output_root_name}"
+        )
+        lines.append("Time stepping (`ROMSRuntimeSettings.time_stepping`):")
+        lines.append(f"- Number of steps (`ntimes`) = {self.time_stepping.ntimes},")
+        lines.append(f"- Time step (`dt`, sec) = {self.time_stepping.dt},")
```

Comment:
should stringification be delegated to the classes this object is composed of to avoid a requirement to update code in two places when the settings classes change?